### PR TITLE
Test3

### PR DESCRIPTION
**Pull request by Gokul-ts**

### CODE REVIEW

File: Training/Training/file.c [hunk 1]
```diff
@@ -7,7 +7,135 @@
...
 void main () {
-	printf ("Hello World");
+   char pieces[] = { ' ',' ',' ',' ',' ',' ',' ',' ',' ' };
+   PrintBoard (pieces);
+   int player1 = 0;
+   int player2 = 0;
+   int input1 = 0, input2 = 0;
+   for (int i = 0; i < 4; i++) {
+      for (;;) {
+         input1 = GetUserInput ("Player1:");
```
Comment (sanjivtmm) [inline Training/Training/file.c]: You should display the board after each player makes a move and also check for win/draw after each move and not after both players move.

File: Training/Training/file.c [hunk 2]
```diff
@@ -7,7 +7,135 @@
...
+         input2 = GetUserInput ("Player2:");
+         if ((input1 > 9 || input1 < 1) || (input2 > 9 || input2 < 1)) {
```
Comment (sanjivtmm) [inline Training/Training/file.c]: No check is being made for repetition of moves.

File: Training/Training/file.c [hunk 3]
```diff
@@ -7,7 +7,135 @@
...
+   return number;
+}
+
+
+void PrintBoard (char pieces[]) {
+   printf ("BOARD GAME\n");
+   printf ("+---+---+---+\n|");
+   for (int j = 0; j < 9; j++) {
+      switch (j) {
+         case 0:
```
Comment (sanjivtmm) [inline Training/Training/file.c]: Not an elegant way to show the board.

File: Training/Training/file.c [hunk 4]
```diff
@@ -7,7 +7,135 @@
...
+            continue;
+         }
+         break;
+      }
+      pieces[input1 - 1] = 'X';
+      pieces[input2 - 1] = 'O';
+      ClrScreen ();
+      PrintBoard (pieces);
+      if (i > 1) {
+         for (int i = 0; i < 1; i++) {
```
Comment (sanjivtmm) [inline Training/Training/file.c]: This is not very elegant.

File: Training/Training/BoardGame.c [hunk 5]
```diff
@@ -0,0 +1,72 @@
+// ------------------------------------------------------------------------------------------------
+// Training ~ A training program for new joiners at Metamation, Batch - July 2024.
+// Copyright Gokul TS (c) Metamation India.
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: DO NOT use your name in the copyright message. You do not own the copyright.
Any copyright is owned by TRUMPF Metamation Private Limited.

File: Training/Training/BoardGame.c [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+// ------------------------------------------------------------------
+// BoardGame.c
+// Program to play board game between two players.
+// ------------------------------------------------------------------------------------------------
+#include <stdio.h>
+#include <conio.h>
+#include <stdlib.h>
+
+void PrintBoard (char pieces[3][3]) {
+   printf ("BOARD GAME\n+---+---+---+\n|");
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: The name of the game is "Tic-Tac-Toe"

File: Training/Training/BoardGame.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+   for (int i = 0; i < 3; i++)
+      for (int j = 0; j < 3; j++) {
+         printf (j == 0 && i == 1 || j == 0 && i == 2 ? "| %c |" : " %c |", pieces[i][j]);
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: The complex condition j == 0 && i == 1 || j == 0 && i == 2 can be reduced to j == 0 && i != 0

File: Training/Training/BoardGame.c [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+#endif 
+}
+
+void GetPlayerInput (char prompt, char* pieces) {
+   for (;;) {
+      printf ("Player %c: ", prompt);
+      char ch = getch ();
+      int input = ch - 49;
+      if (input > 9 || input < 0) {
+         printf ("Please enter a number between 1-9\n");
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: A better prompt is "Please select an empty square between 1 and 9".

File: Training/Training/BoardGame.c [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      if (pieces[input] != 'X' && pieces[input] != 'O') {
+         pieces[input] = prompt;
+         break;
+      } else printf ("Number already chosen!!\n");
+   }
+}
+
+int CheckWin (char pieces[3][3], char currentPlayer) {
+   for (int i = 0; i < 3; i++)
+      if (pieces[i][0] == currentPlayer && pieces[i][1] == currentPlayer && pieces[i][2] == currentPlayer ||
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: Use parentheses to improve readability.

File: Training/Training/BoardGame.c (hunk 7 above)
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: A nice feature would be to show each player's pieces (i.e. O or X) in a different colour.

File: Training/Training/BoardGame.c [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+   printf ("BOARD GAME\n+---+---+---+\n|");
+   for (int i = 0; i < 3; i++)
+      for (int j = 0; j < 3; j++) {
+         printf (j == 0 && i == 1 || j == 0 && i == 2 ? "| %c |" : " %c |", pieces[i][j]);
+         if ((j + 1) % 3 == 0)  printf ("\n+---+---+---+\n");
+      }
+}
+
+void ClrScreen () {
+#ifdef _WIN32
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: Why check for this?

File: Training/Training/BoardGame.c [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+   system ("cls");
+#endif 
+}
+
+void GetPlayerInput (char prompt, char* pieces) {
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: Better names for the parameters are player and board.

File: Training/Training/BoardGame.c [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+         int result = CheckWin (pieces, currentPlayer);
+         if (result) {
+            printf ("Player %c is the Winner!!!\n", currentPlayer);
+            return;
+         } else if (!result && i > 7) {
+            printf ("Match Draw!!!\n");
+            return;
+         }
+      }
+      currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
```
Comment (sanjivtmm) [inline Training/Training/BoardGame.c]: What happens if I replace this line by
currentPlayer ^= ('X' ^ 'O');

How does it work?

**Review by Dakshita-Minocha — APPROVED** [review]

**Pull request closed without merging**